<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="John Zobolas">
<meta name="dcterms.date" content="2023-10-25">
<meta name="description" content="Demonstrate use of survival BART on the lung dataset via mlr3proba and distr6.">

<title>Survival modeling in mlr3 using Bayesian Additive Regression Trees (BART) – mlr-org</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../images/favicon.ico" rel="icon">
<script src="../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-0217fbee799b1e5c46e1a9544cf407f3.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": true,
  "collapse-after": 2,
  "panel-placement": "end",
  "type": "textbox",
  "limit": 5,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">mlr-org</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-overview" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Overview</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-overview">    
        <li>
    <a class="dropdown-item" href="../../../ecosystem.html">
 <span class="dropdown-text">Ecosystem</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../dependencies.html">
 <span class="dropdown-text">Dependencies</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="../../../tasks.html">
 <span class="dropdown-text">Tasks</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../learners.html">
 <span class="dropdown-text">Learners</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../resamplings.html">
 <span class="dropdown-text">Resamplings</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../measures.html">
 <span class="dropdown-text">Measures</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="../../../pipeops.html">
 <span class="dropdown-text">Pipeline Operators</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../graphs.html">
 <span class="dropdown-text">Graphs</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../torch_pipeops.html">
 <span class="dropdown-text">Torch Pipeline Operators</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="../../../tuners.html">
 <span class="dropdown-text">Tuners</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../tuning_spaces.html">
 <span class="dropdown-text">Tuning Spaces</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../terminators.html">
 <span class="dropdown-text">Terminators</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li>
    <a class="dropdown-item" href="../../../filters.html">
 <span class="dropdown-text">Feature Selection Filter</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../fselectors.html">
 <span class="dropdown-text">Feature Selection Wrapper</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../resources.html"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../book.html"> 
<span class="menu-text">Book</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../gallery.html"> 
<span class="menu-text">Gallery</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-benchmarks" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Benchmarks</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-benchmarks">    
        <li>
    <a class="dropdown-item" href="../../../benchmarks/benchmarks_mlr3.html">
 <span class="dropdown-text">mlr3</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../benchmarks/benchmarks_mlr3fselect.html">
 <span class="dropdown-text">mlr3fselect</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../benchmarks/benchmarks_mlr3tuning.html">
 <span class="dropdown-text">mlr3tuning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../benchmarks/benchmarks_async.html">
 <span class="dropdown-text">mlr3tuning async</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../team.html"> 
<span class="menu-text">Team</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-more" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">More</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-more">    
        <li>
    <a class="dropdown-item" href="../../../support.html">
 <span class="dropdown-text">Support</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../contributing.html">
 <span class="dropdown-text">Contributing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../faq.html">
 <span class="dropdown-text">Frequently Asked Questions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../blogroll.html">
 <span class="dropdown-text">Blogroll</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-bi-rss" role="link" data-bs-toggle="dropdown" aria-expanded="false">
      <i class="bi bi-rss" role="img">
</i> 
 <span class="menu-text"></span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-bi-rss">    
        <li>
    <a class="dropdown-item" href="../../../gallery-all.xml">
 <span class="dropdown-text">Gallery RSS</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item compact">
    <a class="nav-link" href="https://lmmisld-lmu-stats-slds.srv.mwn.de/mlr_invite"> <i class="bi bi-chat-dots" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://stackoverflow.com/questions/tagged/mlr3"> <i class="bi bi-stack-overflow" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mlr-org"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Survival modeling in mlr3 using Bayesian Additive Regression Trees (BART)</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
  <div id="quarto-announcement" data-announcement-id="50f0eab22d7d8f5381fa1ea26eadc5d4" class="alert alert-primary hidden"><i class="bi bi-info-circle quarto-announcement-icon"></i><div class="quarto-announcement-content">
<p>The website features runtime and memory <a href="benchmarks/benchmarks_mlr3tuning.qmd">benchmarks</a> of the mlr3tuning package now.</p>
</div><i class="bi bi-x-lg quarto-announcement-action"></i></div>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Survival modeling in mlr3 using Bayesian Additive Regression Trees (BART)</h1>
                  <div>
        <div class="description">
          <p>Demonstrate use of survival BART on the lung dataset via mlr3proba and distr6.</p>
        </div>
      </div>
                </div>
  </div>

  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p><a href="https://github.com/bblodfon">John Zobolas</a> <a href="https://orcid.org/0000-0002-3609-8674" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">October 25, 2023</p>
      </div>
    </div>
    
      
    </div>
    

  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#intro" id="toc-intro" class="nav-link active" data-scroll-target="#intro">Intro</a></li>
  <li><a href="#libraries" id="toc-libraries" class="nav-link" data-scroll-target="#libraries">Libraries</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#train-and-test" id="toc-train-and-test" class="nav-link" data-scroll-target="#train-and-test">Train and Test</a>
  <ul class="collapse">
  <li><a href="#distr" id="toc-distr" class="nav-link" data-scroll-target="#distr">distr</a></li>
  <li><a href="#crank" id="toc-crank" class="nav-link" data-scroll-target="#crank">crank</a></li>
  </ul></li>
  <li><a href="#feature-importance" id="toc-feature-importance" class="nav-link" data-scroll-target="#feature-importance">Feature importance</a></li>
  <li><a href="#mcmc-diagnostics" id="toc-mcmc-diagnostics" class="nav-link" data-scroll-target="#mcmc-diagnostics">MCMC Diagnostics</a></li>
  <li><a href="#performance-test-set" id="toc-performance-test-set" class="nav-link" data-scroll-target="#performance-test-set">Performance (test set)</a></li>
  <li><a href="#resampling" id="toc-resampling" class="nav-link" data-scroll-target="#resampling">Resampling</a></li>
  <li><a href="#uncertainty-quantification-in-survival-prediction" id="toc-uncertainty-quantification-in-survival-prediction" class="nav-link" data-scroll-target="#uncertainty-quantification-in-survival-prediction">Uncertainty Quantification in Survival Prediction</a></li>
  <li><a href="#partial-dependence-plot" id="toc-partial-dependence-plot" class="nav-link" data-scroll-target="#partial-dependence-plot">Partial Dependence Plot</a></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session Information</a>
  <ul class="collapse">
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">




<section id="intro" class="level2">
<h2 class="anchored" data-anchor-id="intro">Intro</h2>
<p>Here are some interesting reads regarding BART:</p>
<ul>
<li>The first BART paper <span class="citation" data-cites="Chipman2010">(<a href="#ref-Chipman2010" role="doc-biblioref">Chipman, George, and McCulloch 2010</a>)</span>.</li>
<li>The first implementation of BART for survival data <span class="citation" data-cites="Bonato2011">(<a href="#ref-Bonato2011" role="doc-biblioref">Bonato et al. 2011</a>)</span>. This includes fully parametric AFT and Weibull models and the semi-parametric CoxPH regression model.</li>
<li>The first non-parametric implementation of BART for survival data <span class="citation" data-cites="Sparapani2016">(<a href="#ref-Sparapani2016" role="doc-biblioref">R. A. Sparapani et al. 2016</a>)</span></li>
<li><code>BART</code> R package tutorial <span class="citation" data-cites="Sparapani2021">(<a href="#ref-Sparapani2021" role="doc-biblioref">R. Sparapani, Spanbauer, and McCulloch 2021</a>)</span></li>
</ul>
<p>We incorporated the survival <code>BART</code> model in <code>mlr3extralearners</code> and in this tutorial we will demonstrate how we can use packages like <code>mlr3</code>, <code>mlr3proba</code> and <code>distr6</code> to more easily manipulate the output predictions to assess model convergence, validate our model (via several survival metrics), as well as perform model interpretation via PDPs (Partial Dependence Plots).</p>
</section>
<section id="libraries" class="level2">
<h2 class="anchored" data-anchor-id="libraries">Libraries</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3extralearners)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3pipelines)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3proba)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(distr6)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BART) <span class="co"># 2.9.4</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<p>We will use the Lung Cancer Dataset. We convert the <code>time</code> variable from days to months to ease the computational burden:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>task_lung <span class="ot">=</span> <span class="fu">tsk</span>(<span class="st">'lung'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> task_lung<span class="sc">$</span><span class="fu">data</span>()</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># in case we want to select specific columns to keep</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># d = d[ ,colnames(d) %in% c("time", "status", "age", "sex", "ph.karno"), with = FALSE]</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>time <span class="ot">=</span> <span class="fu">ceiling</span>(d<span class="sc">$</span>time<span class="sc">/</span><span class="fl">30.44</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>task_lung <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung'</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>task_lung<span class="sc">$</span>label <span class="ot">=</span> <span class="st">"Lung Cancer"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>The original <code>BART</code> implementation supports categorical features (factors). This results in different importance scores per each dummy level which doesn’t work well with <code>mlr3</code>. So features of type <code>factor</code> or <code>character</code> are not allowed and we leave it to the user to encode them as they please.</li>
<li>The original <code>BART</code> implementation supports features with missing values. This is totally fine with <code>mlr3</code> as well! In this example, we impute the features to show good ML practice.</li>
</ol>
</div>
</div>
<p>In our lung dataset, we encode the <code>sex</code> feature and perform model-based imputation with the <code>rpart</code> regression learner:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>po_encode <span class="ot">=</span> <span class="fu">po</span>(<span class="st">'encode'</span>, <span class="at">method =</span> <span class="st">'treatment'</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>po_impute <span class="ot">=</span> <span class="fu">po</span>(<span class="st">'imputelearner'</span>, <span class="fu">lrn</span>(<span class="st">'regr.rpart'</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>pre <span class="ot">=</span> po_encode <span class="sc">%&gt;&gt;%</span> po_impute</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>task <span class="ot">=</span> pre<span class="sc">$</span><span class="fu">train</span>(task_lung)[[<span class="dv">1</span>]]</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>task</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TaskSurv:lung&gt; (168 x 9): Lung Cancer
* Target: time, status
* Properties: -
* Features (7):
  - int (6): age, meal.cal, pat.karno, ph.ecog, ph.karno, wt.loss
  - dbl (1): sex</code></pre>
</div>
</div>
<p>No missing values in our data:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span><span class="fu">missings</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     time    status       age  meal.cal pat.karno   ph.ecog  ph.karno   wt.loss       sex 
        0         0         0         0         0         0         0         0         0 </code></pre>
</div>
</div>
<p>We partition the data to train and test sets:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>part <span class="ot">=</span> <span class="fu">partition</span>(task, <span class="at">ratio =</span> <span class="fl">0.9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="train-and-test" class="level2">
<h2 class="anchored" data-anchor-id="train-and-test">Train and Test</h2>
<p>We train the <code>BART</code> model and predict on the test set:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default `ndpost` value: 1000. We reduce it to 50 to speed up calculations in this tutorial</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>learner <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.bart"</span>, <span class="at">nskip =</span> <span class="dv">250</span>, <span class="at">ndpost =</span> <span class="dv">50</span>, <span class="at">keepevery =</span> <span class="dv">10</span>, <span class="at">mc.cores =</span> <span class="dv">10</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span><span class="fu">train</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>test)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See more details about <code>BART</code>’s parameters on the online <a href="https://mlr3extralearners.mlr-org.com/reference/mlr_learners_surv.bart.html">documentation</a>.</p>
<section id="distr" class="level3">
<h3 class="anchored" data-anchor-id="distr">distr</h3>
<p>What kind of object is the predicted <code>distr</code>?</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>distr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Arrdist(17x30x50) </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Arrdist dimensions:
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Patients (observations)</li>
<li>Time points (months)</li>
<li>Number of posterior draws</li>
</ol>
</div>
</div>
<p>Actually the <code>$distr</code> is an active <a href="https://r6.r-lib.org/index.html">R6</a> field - this means that some computation is required to create it. What the prediction object actually stores internally is a 3d survival array (can be used directly with no performance overhead):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(p<span class="sc">$</span>data<span class="sc">$</span>distr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 17 30 50</code></pre>
</div>
</div>
<p>This is a more easy-to-understand and manipulate form of the full posterior survival matrix prediction from the <code>BART</code> package (<span class="citation" data-cites="Sparapani2021">(<a href="#ref-Sparapani2021" role="doc-biblioref">R. Sparapani, Spanbauer, and McCulloch 2021</a>)</span>, pages 34-35).</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Though we have optimized with C++ code the way the <code>Arrdist</code> object is constructed, calling the <code>$distr</code> field can be computationally taxing if the product of the sizes of the 3 dimensions above <strong>exceeds ~1 million</strong>. In our case, <span class="math inline">\(23 \times 31 \times 50 = 35650\)</span> so the conversion to an <code>Arrdist</code> via <code>$distr</code> will certainly not create performance issues.</p>
</div>
</div>
<p>An example using the internal prediction data: get all the posterior probabilities of the 3rd patient in the test set, at 12 months (1 year):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>data<span class="sc">$</span>distr[<span class="dv">3</span>, <span class="dv">12</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 0.16682639 0.52846967 0.51057138 0.23143445 0.24949314 0.23818569 0.28702429 0.39494813 0.68549420 0.38617165
[11] 0.18528656 0.18365059 0.58371359 0.34762244 0.30502214 0.20621051 0.73117066 0.37176522 0.44852366 0.48714300
[21] 0.13227459 0.36120922 0.04110908 0.08701728 0.26405672 0.13075270 0.33063856 0.71372303 0.28818814 0.43886058
[31] 0.16152343 0.11132334 0.09516388 0.26125352 0.16578149 0.33411388 0.32607511 0.43967968 0.25908991 0.21763759
[41] 0.47965258 0.50324377 0.13649939 0.40579983 0.29247802 0.09903617 0.56096633 0.26853225 0.20253541 0.42479851</code></pre>
</div>
</div>
<p>Working with the <code>$distr</code> interface and <code>Arrdist</code> objects is very efficient as we will see later for predicting survival estimates.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>In survival analysis, <span class="math inline">\(S(t) = 1 - F(t)\)</span>, where <span class="math inline">\(S(t)\)</span> the survival function and <span class="math inline">\(F(t)\)</span> the cumulative distribution function (cdf). The latter can be interpreted as <code>risk</code> or probability of death up to time <span class="math inline">\(t\)</span>.</p>
<p>We can verify the above from the prediction object:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>surv_array <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> distr6<span class="sc">::</span><span class="fu">gprm</span>(p<span class="sc">$</span>distr, <span class="st">"cdf"</span>) <span class="co"># 3d array</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">all_equal</span>(p<span class="sc">$</span>data<span class="sc">$</span>distr, surv_array)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `all_equal()` was deprecated in dplyr 1.1.0.
ℹ Please use `all.equal()` instead.
ℹ And manually order the rows/cols as needed</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>`y` must be a data frame.</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="crank" class="level3">
<h3 class="anchored" data-anchor-id="crank">crank</h3>
<p><code>crank</code> is the <strong>expected mortality</strong> <span class="citation" data-cites="Sonabend2022">(<a href="#ref-Sonabend2022" role="doc-biblioref">Sonabend, Bender, and Vollmer 2022</a>)</span> which is the sum of the predicted cumulative hazard function (as is done in random survival forest models). Higher values denote larger risk. To calculate <code>crank</code>, we need a survival matrix. So we have to choose which 3rd dimension we should use from the predicted survival array. This is what the <code>which.curve</code> parameter of the <code>learner</code> does:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span>param_set<span class="sc">$</span><span class="fu">get_values</span>()<span class="sc">$</span>which.curve</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5</code></pre>
</div>
</div>
<p>The default value (<span class="math inline">\(0.5\)</span> quantile) is the <strong>median survival probability</strong>. It could be any other quantile (e.g.&nbsp;<span class="math inline">\(0.25\)</span>). Other possible values for <code>which.curve</code> are <code>mean</code> or a number denoting the exact posterior draw to extract (e.g.&nbsp;the last one, <code>which.curve = 50</code>).</p>
</section>
</section>
<section id="feature-importance" class="level2">
<h2 class="anchored" data-anchor-id="feature-importance">Feature importance</h2>
<p>Default score is the <strong>observed count of each feature</strong> in the trees (so the higher the score, the more important the feature):</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span>param_set<span class="sc">$</span>values<span class="sc">$</span>importance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "count"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span><span class="fu">importance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      sex   wt.loss  ph.karno  meal.cal pat.karno       age   ph.ecog 
     8.94      8.22      8.16      7.74      6.44      6.28      6.28 </code></pre>
</div>
</div>
</section>
<section id="mcmc-diagnostics" class="level2">
<h2 class="anchored" data-anchor-id="mcmc-diagnostics">MCMC Diagnostics</h2>
<p><code>BART</code> uses internally MCMC (Markov Chain Monte Carlo) to sample from the posterior survival distribution. We need to check that MCMC has converged, meaning that the chains have reached a stationary distribution that approximates the true posterior survival distribution (otherwise the predictions may be inaccurate, misleading and unreliable).</p>
<p>We use Geweke’s convergence diagnostic test as it is implemented in the <code>BART</code> R package. We choose 10 random patients from the train set to evaluate the MCMC convergence.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># predictions on the train set</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>p_train <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># choose 10 patients from the train set randomly and make a list</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">=</span> <span class="fu">as.list</span>(<span class="fu">sample</span>(<span class="fu">length</span>(part<span class="sc">$</span>train), <span class="dv">10</span>))</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>z_list <span class="ot">=</span> <span class="fu">lapply</span>(ids, <span class="cf">function</span>(id) {</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># matrix with columns =&gt; time points and rows =&gt; posterior draws</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  post_surv <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">t</span>(distr6<span class="sc">::</span><span class="fu">gprm</span>(p_train<span class="sc">$</span>distr[id], <span class="st">"cdf"</span>)[<span class="dv">1</span>,,])</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  BART<span class="sc">::</span><span class="fu">gewekediag</span>(post_surv)<span class="sc">$</span>z <span class="co"># get the z-scores</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the z scores vs time for all patients</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(z_list) <span class="sc">%&gt;%</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">name =</span> <span class="fu">as.numeric</span>(name)) <span class="sc">%&gt;%</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> name, <span class="at">y =</span> value)) <span class="sc">+</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Time (months)"</span>, <span class="at">y =</span> <span class="st">"Z-scores"</span>) <span class="sc">+</span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add critical values for a = 0.05</span></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">1.96</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fl">1.96</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/diag-mcmc-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Geweke plot for MCMC diagnostics. Z-scores for the difference in the mean survival prediction between the first 10% and last 50% part of a Markov chain. The predictions are taken from 10 random patients in the train set. Red lines indicate the a = 0.05 critical line. Only a few z-scores exceed the 95% limits so we conclude that convergence has been attained.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="performance-test-set" class="level2">
<h2 class="anchored" data-anchor-id="performance-test-set">Performance (test set)</h2>
<p>We will use the following survival metrics:</p>
<ol type="1">
<li>Integrated Brier Score (requires a survival distribution prediction - <code>distr</code>)</li>
<li>Uno’s C-index (requires a continuous ranking score prediction - <code>crank</code>)</li>
</ol>
<p>For the first measure we will use the ERV (<strong>Explained Residual Variation</strong>) version, which standardizes the score against a Kaplan-Meier (KM) baseline <span class="citation" data-cites="Sonabend2022a">(<a href="#ref-Sonabend2022a" role="doc-biblioref">Sonabend et al. 2022</a>)</span>. This means that values close to <span class="math inline">\(0\)</span> represent performance similar to a KM model, negative values denote worse performance than KM and <span class="math inline">\(1\)</span> is the absolute best possible score.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>measures <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msr</span>(<span class="st">"surv.graf"</span>, <span class="at">ERV =</span> <span class="cn">TRUE</span>),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">weight_meth =</span> <span class="st">"G2"</span>, <span class="at">id =</span> <span class="st">"surv.cindex.uno"</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (measure <span class="cf">in</span> measures) {</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p<span class="sc">$</span><span class="fu">score</span>(measure, <span class="at">task =</span> task, <span class="at">train_set =</span> part<span class="sc">$</span>train))</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>surv.graf 
0.1607549 
surv.cindex.uno 
      0.7120706 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>All metrics use by default the <strong>median survival distribution</strong> from the 3d array, no matter what is the <code>which.curve</code> argument during the learner’s construction.</p>
</div>
</div>
</section>
<section id="resampling" class="level2">
<h2 class="anchored" data-anchor-id="resampling">Resampling</h2>
<p>Performing resampling with the <code>BART</code> learner is very easy using <code>mlr3</code>.</p>
<p>We first stratify the data by <code>status</code>, so that in each resampling the proportion of censored vs un-censored patients remains the same:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span>col_roles<span class="sc">$</span>stratum <span class="ot">=</span> <span class="st">'status'</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span>strata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       N                row_id
   &lt;int&gt;                &lt;list&gt;
1:   121       1,2,4,5,6,7,...
2:    47  3,26,49,52,62,64,...</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>rr <span class="ot">=</span> <span class="fu">resample</span>(task, learner, <span class="at">resampling =</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">5</span>), <span class="at">store_backends =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [14:22:15.342] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 1/5)
INFO  [14:22:16.982] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 2/5)
INFO  [14:22:18.456] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 3/5)
INFO  [14:22:19.907] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 4/5)
INFO  [14:22:21.351] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 5/5)</code></pre>
</div>
</div>
<p>No errors or warnings:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span>errors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Empty data.table (0 rows and 2 cols): iteration,msg</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span>warnings</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Empty data.table (0 rows and 2 cols): iteration,msg</code></pre>
</div>
</div>
<p>Performance in each fold:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   task_id learner_id resampling_id iteration   surv.graf surv.cindex.uno
    &lt;char&gt;     &lt;char&gt;        &lt;char&gt;     &lt;int&gt;       &lt;num&gt;           &lt;num&gt;
1:    lung  surv.bart            cv         1  0.06674423       0.6270652
2:    lung  surv.bart            cv         2 -0.06994762       0.5789767
3:    lung  surv.bart            cv         3 -0.05637324       0.5880529
4:    lung  surv.bart            cv         4  0.02718982       0.6126352
5:    lung  surv.bart            cv         5 -0.08941180       0.5695597
Hidden columns: task, learner, resampling, prediction_test</code></pre>
</div>
</div>
<p>Mean cross-validation performance:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      surv.graf surv.cindex.uno 
    -0.02435972      0.59525794 </code></pre>
</div>
</div>
</section>
<section id="uncertainty-quantification-in-survival-prediction" class="level2">
<h2 class="anchored" data-anchor-id="uncertainty-quantification-in-survival-prediction">Uncertainty Quantification in Survival Prediction</h2>
<p>We will choose two patients from the test set and plot their survival prediction posterior estimates.</p>
<p>Let’s choose the patients with the worst and the best survival time:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>death_times <span class="ot">=</span> p<span class="sc">$</span>truth[,<span class="dv">1</span>]</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(death_times)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  2  4  6  6  6  7  8  9 10 10 10 10 13 15 15 19 22</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>worst_indx <span class="ot">=</span> <span class="fu">which</span>(death_times <span class="sc">==</span> <span class="fu">min</span>(death_times))[<span class="dv">1</span>] <span class="co"># died first</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>best_indx  <span class="ot">=</span> <span class="fu">which</span>(death_times <span class="sc">==</span> <span class="fu">max</span>(death_times))[<span class="dv">1</span>] <span class="co"># died last</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>patient_ids <span class="ot">=</span> <span class="fu">c</span>(worst_indx, best_indx)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>patient_ids <span class="co"># which patient IDs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12  9</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>death_times <span class="ot">=</span> death_times[patient_ids]</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>death_times <span class="co"># 1st is worst, 2nd is best</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  2 22</code></pre>
</div>
</div>
<p>Subset <code>Arrdist</code> to only the above 2 patients:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>arrd <span class="ot">=</span> p<span class="sc">$</span>distr[patient_ids]</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>arrd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Arrdist(2x30x50) </code></pre>
</div>
</div>
<p>We choose time points (in months) for the survival estimates:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>months <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">36</span>) <span class="co"># 1 month - 3 years</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use the <code>$distr</code> interface and the <code>$survival</code> property to get survival probabilities from an <code>Arrdist</code> object as well as the <strong>quantile credible intervals</strong> (CIs). The median survival probabilities can be extracted as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>med <span class="ot">=</span> arrd<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 'med' for median</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(med) <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_med"</span>)</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>med <span class="ot">=</span> <span class="fu">as_tibble</span>(med) <span class="sc">%&gt;%</span> <span class="fu">add_column</span>(<span class="at">month =</span> months)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(med)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  `12_med` `9_med` month
     &lt;dbl&gt;   &lt;dbl&gt; &lt;int&gt;
1    0.857   0.974     1
2    0.741   0.948     2
3    0.630   0.923     3
4    0.524   0.900     4
5    0.446   0.872     5
6    0.362   0.833     6</code></pre>
</div>
</div>
<p>We can briefly verify model’s predictions: 1st patient survival probabilities on any month are lower (worst) compared to the 2nd patient.</p>
<p>Note that subsetting an <code>Arrdist</code> (3d array) creates a <code>Matdist</code> (2d matrix), for example we can explicitly get the median survival probabilities:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>matd_median <span class="ot">=</span> arrd[, <span class="fl">0.5</span>] <span class="co"># median</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_median<span class="sc">$</span><span class="fu">survival</span>(months)) <span class="co"># same as with `arrd`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.8574541 0.9736144
2 0.7409911 0.9479014
3 0.6304261 0.9226185
4 0.5236389 0.8996011
5 0.4457717 0.8721516
6 0.3620130 0.8329688</code></pre>
</div>
</div>
<p>Using the <code>mean</code> posterior survival probabilities or the ones from the last posterior draw is also possible and can be done as follows:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>matd_mean <span class="ot">=</span> arrd[, <span class="st">"mean"</span>] <span class="co"># mean (if needed)</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_mean<span class="sc">$</span><span class="fu">survival</span>(months))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.8494062 0.9721130
2 0.7246768 0.9445261
3 0.6213767 0.9154688
4 0.5246258 0.8865332
5 0.4487428 0.8576749
6 0.3644190 0.8155328</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>matd_50draw <span class="ot">=</span> arrd[, <span class="dv">50</span>] <span class="co"># the 50th posterior draw</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_50draw<span class="sc">$</span><span class="fu">survival</span>(months))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.7671943 0.9865953
2 0.5885871 0.9733703
3 0.4515607 0.9603225
4 0.3464348 0.9474497
5 0.2657828 0.9347494
6 0.1812379 0.9112550</code></pre>
</div>
</div>
<p>To get the CIs we will subset the <code>Arrdist</code> using a quantile number (0-1), which extracts a <code>Matdist</code> based on the cdf. The survival function is 1 - cdf, so low and upper bounds are reversed:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>low  <span class="ot">=</span> arrd[, <span class="fl">0.975</span>]<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 2.5% bound</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>high <span class="ot">=</span> arrd[, <span class="fl">0.025</span>]<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 97.5% bound</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(low)  <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_low"</span>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(high) <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_high"</span>)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>low  <span class="ot">=</span> <span class="fu">as_tibble</span>(low)</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>high <span class="ot">=</span> <span class="fu">as_tibble</span>(high)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The median posterior survival probabilities for the two patient of interest and the corresponding CI bounds in a tidy format are:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>surv_tbl <span class="ot">=</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(low, med, high) <span class="sc">%&gt;%</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>month, <span class="at">values_to =</span> <span class="st">"surv"</span>,</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"patient_id"</span>, <span class="st">".value"</span>), <span class="at">names_sep =</span> <span class="st">"_"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(patient_id)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>surv_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 72 × 5
   patient_id month    low   med  high
   &lt;chr&gt;      &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 12             1 0.639  0.857 0.962
 2 9              1 0.929  0.974 0.994
 3 12             2 0.398  0.741 0.925
 4 9              2 0.864  0.948 0.988
 5 12             3 0.230  0.630 0.889
 6 9              3 0.803  0.923 0.980
 7 12             4 0.133  0.524 0.828
 8 9              4 0.761  0.900 0.976
 9 12             5 0.0773 0.446 0.770
10 9              5 0.722  0.872 0.971
# ℹ 62 more rows</code></pre>
</div>
</div>
<p>We draw survival curves with the uncertainty for the survival probability quantified:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>my_colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#E41A1C"</span>, <span class="st">"#4DAF4A"</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(my_colors) <span class="ot">=</span> patient_ids</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>surv_tbl <span class="sc">%&gt;%</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> month, <span class="at">y =</span> med)) <span class="sc">+</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="fu">aes</span>(<span class="at">color =</span> patient_id), <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'Time (Months)'</span>) <span class="sc">+</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Survival Probability'</span>) <span class="sc">+</span></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high, <span class="at">fill =</span> patient_id),</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> death_times[<span class="dv">1</span>], <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> my_colors[<span class="dv">1</span>]) <span class="sc">+</span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> death_times[<span class="dv">2</span>], <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> my_colors[<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb59-15"><a href="#cb59-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb59-16"><a href="#cb59-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">"Patient ID"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/surv-with-uq-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Uncertainty quantification for the survival prediction of two patients in the test set using 95% credible intervals. The two vertical lines correspond to the reported time of death (in months) for the two patients.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="partial-dependence-plot" class="level2">
<h2 class="anchored" data-anchor-id="partial-dependence-plot">Partial Dependence Plot</h2>
<p>We will use a Partial Dependence Plot (PDP) <span class="citation" data-cites="Friedman2001">(<a href="#ref-Friedman2001" role="doc-biblioref">Friedman 2001</a>)</span> to visualize how much different are males vs females in terms of their average survival predictions across time.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>PDPs assume that features are independent. In our case we need to check that <code>sex</code> doesn’t correlate with any of the other features used for training the <code>BART</code> learner. Since <code>sex</code> is a categorical feature, we fit a linear model using as target variable every other feature in the data (<span class="math inline">\(lm(feature \sim sex)\)</span>) and conduct an ANOVA (ANalysis Of VAriance) to get the variance explained or <span class="math inline">\(R^2\)</span>. The square root of that value is the correlation measure we want.</p>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># code from https://christophm.github.io/interpretable-ml-book/ale.html</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>mycor <span class="ot">=</span> <span class="cf">function</span>(cnames, data) {</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  x.num <span class="ot">=</span> data[, cnames[<span class="dv">1</span>], with <span class="ot">=</span> <span class="cn">FALSE</span>][[<span class="dv">1</span>]]</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  x.cat <span class="ot">=</span> data[, cnames[<span class="dv">2</span>], with <span class="ot">=</span> <span class="cn">FALSE</span>][[<span class="dv">1</span>]]</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># R^2 = Cor(X, Y)^2 in simple linear regression</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sqrt</span>(<span class="fu">summary</span>(<span class="fu">lm</span>(x.num <span class="sc">~</span> x.cat))<span class="sc">$</span>r.squared)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>cnames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sex"</span>)</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>combs <span class="ot">=</span> <span class="fu">expand.grid</span>(<span class="at">y =</span> <span class="fu">setdiff</span>(<span class="fu">colnames</span>(d), <span class="st">"sex"</span>), <span class="at">x =</span> cnames)</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>combs<span class="sc">$</span>cor <span class="ot">=</span> <span class="fu">apply</span>(combs, <span class="dv">1</span>, mycor, <span class="at">data =</span> task<span class="sc">$</span><span class="fu">data</span>()) <span class="co"># use the train set</span></span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>combs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          y   x        cor
1      time sex 0.11338256
2    status sex 0.22105006
3       age sex 0.12814230
4  meal.cal sex 0.16813658
5 pat.karno sex 0.07075345
6   ph.ecog sex 0.01197613
7  ph.karno sex 0.01512325
8   wt.loss sex 0.17288403</code></pre>
</div>
</div>
<p><code>sex</code> doesn’t correlate strongly with any other feature, so we can compute the PDP:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create two datasets: one with males and one with females</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all other features remain the same (use train data, 205 patients)</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> task<span class="sc">$</span><span class="fu">data</span>(<span class="at">rows =</span> part<span class="sc">$</span>train) <span class="co"># `rows = part$test` to use the test set</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>sex <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>task_males <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung-males'</span>)</span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>sex <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>task_females <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung-females'</span>)</span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make predictions</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>p_males   <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task_males)</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>p_females <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task_females)</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a><span class="co"># take the median posterior survival probability</span></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>surv_males   <span class="ot">=</span> p_males<span class="sc">$</span>distr<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># patients x times</span></span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a>surv_females <span class="ot">=</span> p_females<span class="sc">$</span>distr<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># patients x times</span></span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a><span class="co"># tidy up data: average and quantiles across patients</span></span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a>data_males <span class="ot">=</span></span>
<span id="cb62-20"><a href="#cb62-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(surv_males, <span class="dv">1</span>, <span class="cf">function</span>(row) {</span>
<span id="cb62-21"><a href="#cb62-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb62-22"><a href="#cb62-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">low =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.025</span>),</span>
<span id="cb62-23"><a href="#cb62-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">avg =</span> <span class="fu">mean</span>(row),</span>
<span id="cb62-24"><a href="#cb62-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">high =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.975</span>)</span>
<span id="cb62-25"><a href="#cb62-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb62-26"><a href="#cb62-26" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb62-27"><a href="#cb62-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb62-28"><a href="#cb62-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_column</span>(<span class="at">sex =</span> <span class="st">'male'</span>, <span class="at">month =</span> months, <span class="at">.before =</span> <span class="dv">1</span>)</span>
<span id="cb62-29"><a href="#cb62-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-30"><a href="#cb62-30" aria-hidden="true" tabindex="-1"></a>data_females <span class="ot">=</span></span>
<span id="cb62-31"><a href="#cb62-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(surv_females, <span class="dv">1</span>, <span class="cf">function</span>(row) {</span>
<span id="cb62-32"><a href="#cb62-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb62-33"><a href="#cb62-33" aria-hidden="true" tabindex="-1"></a>      <span class="at">low =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.025</span>),</span>
<span id="cb62-34"><a href="#cb62-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">avg =</span> <span class="fu">mean</span>(row),</span>
<span id="cb62-35"><a href="#cb62-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">high =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.975</span>)</span>
<span id="cb62-36"><a href="#cb62-36" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb62-37"><a href="#cb62-37" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb62-38"><a href="#cb62-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span></span>
<span id="cb62-39"><a href="#cb62-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_column</span>(<span class="at">sex =</span> <span class="st">'female'</span>, <span class="at">month =</span> months, <span class="at">.before =</span> <span class="dv">1</span>)</span>
<span id="cb62-40"><a href="#cb62-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-41"><a href="#cb62-41" aria-hidden="true" tabindex="-1"></a>pdp_tbl <span class="ot">=</span> <span class="fu">bind_rows</span>(data_males, data_females)</span>
<span id="cb62-42"><a href="#cb62-42" aria-hidden="true" tabindex="-1"></a>pdp_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 72 × 5
   sex   month    low   avg  high
   &lt;chr&gt; &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 male      1 0.809  0.944 0.979
 2 male      2 0.652  0.892 0.958
 3 male      3 0.525  0.842 0.934
 4 male      4 0.411  0.792 0.911
 5 male      5 0.332  0.748 0.889
 6 male      6 0.240  0.686 0.851
 7 male      7 0.168  0.635 0.816
 8 male      8 0.120  0.579 0.775
 9 male      9 0.0777 0.523 0.728
10 male     10 0.0504 0.467 0.687
# ℹ 62 more rows</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>my_colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#E41A1C"</span>, <span class="st">"#4DAF4A"</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(my_colors) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'male'</span>, <span class="st">'female'</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>pdp_tbl <span class="sc">%&gt;%</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> month, <span class="at">y =</span> avg)) <span class="sc">+</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="fu">aes</span>(<span class="at">color =</span> sex), <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'Time (Months)'</span>) <span class="sc">+</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Survival Probability'</span>) <span class="sc">+</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high, <span class="at">fill =</span> sex), <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/surv-pdp-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Friedman’s partial dependence function with 95% prediction intervals: males vs females. Females show on average larger survival estimates compared to men, across all time points. Overlapping shaded area represents men and women that have similar survival characteristics.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="session-information" class="level1">
<h1>Session Information</h1>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>(<span class="at">info =</span> <span class="st">"packages"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>═ Session info ═══════════════════════════════════════════════════════════════════════════════════════════════════════
─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
 ! package           * version    date (UTC) lib source
   backports           1.5.0      2024-05-23 [1] CRAN (R 4.4.1)
   BART              * 2.9.9      2024-06-21 [1] CRAN (R 4.4.1)
   checkmate           2.3.2      2024-07-29 [1] CRAN (R 4.4.1)
   cli                 3.6.3      2024-06-21 [1] CRAN (R 4.4.1)
 P codetools           0.2-20     2024-03-31 [?] CRAN (R 4.4.0)
   colorspace          2.1-1      2024-07-26 [1] CRAN (R 4.4.1)
   crayon              1.5.3      2024-06-20 [1] CRAN (R 4.4.1)
   data.table        * 1.16.2     2024-10-10 [1] CRAN (R 4.4.1)
   dictionar6          0.1.3      2021-09-13 [1] CRAN (R 4.4.1)
   digest              0.6.37     2024-08-19 [1] CRAN (R 4.4.1)
   distr6            * 1.8.4      2024-10-17 [1] Github (xoopR/distr6@a642cd3)
   dplyr             * 1.1.4      2023-11-17 [1] CRAN (R 4.4.1)
   evaluate            1.0.1      2024-10-10 [1] CRAN (R 4.4.1)
   fansi               1.0.6      2023-12-08 [1] CRAN (R 4.4.1)
   farver              2.1.2      2024-05-13 [1] CRAN (R 4.4.1)
   fastmap             1.2.0      2024-05-15 [1] CRAN (R 4.4.1)
   future              1.34.0     2024-07-29 [1] CRAN (R 4.4.1)
   future.apply        1.11.2     2024-03-28 [1] CRAN (R 4.4.1)
   generics            0.1.3      2022-07-05 [1] CRAN (R 4.4.1)
   ggplot2           * 3.5.1      2024-04-23 [1] CRAN (R 4.4.1)
   globals             0.16.3     2024-03-08 [1] CRAN (R 4.4.1)
   glue                1.8.0      2024-09-30 [1] CRAN (R 4.4.1)
   gtable              0.3.5      2024-04-22 [1] CRAN (R 4.4.1)
   htmltools           0.5.8.1    2024-04-04 [1] CRAN (R 4.4.1)
   htmlwidgets         1.6.4      2023-12-06 [1] CRAN (R 4.4.1)
   jsonlite            1.8.9      2024-09-20 [1] CRAN (R 4.4.1)
   knitr               1.48       2024-07-07 [1] CRAN (R 4.4.1)
   labeling            0.4.3      2023-08-29 [1] CRAN (R 4.4.1)
 P lattice             0.22-5     2023-10-24 [?] CRAN (R 4.3.3)
   lgr                 0.4.4      2022-09-05 [1] CRAN (R 4.4.1)
   lifecycle           1.0.4      2023-11-07 [1] CRAN (R 4.4.1)
   listenv             0.9.1      2024-01-29 [1] CRAN (R 4.4.1)
   magrittr            2.0.3      2022-03-30 [1] CRAN (R 4.4.1)
 P Matrix              1.7-0      2024-04-26 [?] CRAN (R 4.4.0)
   mlr3              * 0.21.0     2024-09-24 [1] CRAN (R 4.4.1)
   mlr3extralearners * 0.9.0-9000 2024-10-18 [1] Github (mlr-org/mlr3extralearners@a622524)
   mlr3misc            0.15.1     2024-06-24 [1] CRAN (R 4.4.1)
   mlr3pipelines     * 0.7.0      2024-09-24 [1] CRAN (R 4.4.1)
   mlr3proba         * 0.6.9      2024-10-17 [1] Github (mlr-org/mlr3proba@45a5b79)
   mlr3viz             0.9.0      2024-07-01 [1] CRAN (R 4.4.1)
   mlr3website       * 0.0.0.9000 2024-10-18 [1] Github (mlr-org/mlr3website@20d1ddf)
   munsell             0.5.1      2024-04-01 [1] CRAN (R 4.4.1)
 P nlme              * 3.1-165    2024-06-06 [?] CRAN (R 4.4.0)
 P nnet              * 7.3-19     2023-05-03 [?] CRAN (R 4.3.3)
   ooplah              0.2.0      2022-01-21 [1] CRAN (R 4.4.1)
   palmerpenguins      0.1.1      2022-08-15 [1] CRAN (R 4.4.1)
   paradox             1.0.1      2024-07-09 [1] CRAN (R 4.4.1)
   parallelly          1.38.0     2024-07-27 [1] CRAN (R 4.4.1)
   param6              0.2.4      2024-10-17 [1] Github (xoopR/param6@0fa3577)
   pillar              1.9.0      2023-03-22 [1] CRAN (R 4.4.1)
   pkgconfig           2.0.3      2019-09-22 [1] CRAN (R 4.4.1)
   purrr               1.0.2      2023-08-10 [1] CRAN (R 4.4.1)
   R6                  2.5.1      2021-08-19 [1] CRAN (R 4.4.1)
   Rcpp                1.0.13     2024-07-17 [1] CRAN (R 4.4.1)
   renv                1.0.11     2024-10-12 [1] CRAN (R 4.4.1)
   rlang               1.1.4      2024-06-04 [1] CRAN (R 4.4.1)
   rmarkdown           2.28       2024-08-17 [1] CRAN (R 4.4.1)
 P rpart               4.1.23     2023-12-05 [?] CRAN (R 4.4.0)
   scales              1.3.0      2023-11-28 [1] CRAN (R 4.4.1)
   sessioninfo         1.2.2      2021-12-06 [1] CRAN (R 4.4.1)
   set6                0.2.6      2024-10-17 [1] Github (xoopR/set6@a901255)
 P survival          * 3.7-0      2024-06-05 [?] CRAN (R 4.4.0)
   tibble            * 3.2.1      2023-03-20 [1] CRAN (R 4.4.1)
   tidyr             * 1.3.1      2024-01-24 [1] CRAN (R 4.4.1)
   tidyselect          1.2.1      2024-03-11 [1] CRAN (R 4.4.1)
   utf8                1.2.4      2023-10-22 [1] CRAN (R 4.4.1)
   uuid                1.2-1      2024-07-29 [1] CRAN (R 4.4.1)
   vctrs               0.6.5      2023-12-01 [1] CRAN (R 4.4.1)
   withr               3.0.1      2024-07-31 [1] CRAN (R 4.4.1)
   xfun                0.48       2024-10-03 [1] CRAN (R 4.4.1)
   yaml                2.3.10     2024-07-26 [1] CRAN (R 4.4.1)

 [1] /home/marc/repositories/mlr3website/mlr-org/renv/library/linux-ubuntu-noble/R-4.4/x86_64-pc-linux-gnu
 [2] /home/marc/.cache/R/renv/sandbox/linux-ubuntu-noble/R-4.4/x86_64-pc-linux-gnu/9a444a72

 P ── Loaded and on-disk path mismatch.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
<section id="references" class="level2">




</section>
</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Bonato2011" class="csl-entry" role="listitem">
Bonato, Vinicius, Veerabhadran Baladandayuthapani, Bradley M. Broom, Erik P. Sulman, Kenneth D. Aldape, and Kim Anh Do. 2011. <span>“<span class="nocase">Bayesian ensemble methods for survival prediction in gene expression data</span>.”</span> <em>Bioinformatics</em> 27 (3): 359–67. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTQ660">https://doi.org/10.1093/BIOINFORMATICS/BTQ660</a>.
</div>
<div id="ref-Chipman2010" class="csl-entry" role="listitem">
Chipman, Hugh A, Edward I George, and Robert E McCulloch. 2010. <span>“<span>BART: BAYESIAN ADDITIVE REGRESSION TREES</span>.”</span> <em>The Annals of Applied Statistics</em> 4 (1): 266–98. <a href="http://www.jstor.org/stable/27801587">http://www.jstor.org/stable/27801587</a>.
</div>
<div id="ref-Friedman2001" class="csl-entry" role="listitem">
Friedman, Jerome H. 2001. <span>“<span class="nocase">Greedy function approximation: a gradient boosting machine</span>.”</span> <em>Annals of Statistics</em>, 1189–1232. <a href="https://doi.org/10.1214/aos/1013203451">https://doi.org/10.1214/aos/1013203451</a>.
</div>
<div id="ref-Sonabend2022" class="csl-entry" role="listitem">
Sonabend, Raphael, Andreas Bender, and Sebastian Vollmer. 2022. <span>“<span class="nocase">Avoiding C-hacking when evaluating survival distribution predictions with discrimination measures</span>.”</span> Edited by Zhiyong Lu. <em>Bioinformatics</em>, July. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTAC451">https://doi.org/10.1093/BIOINFORMATICS/BTAC451</a>.
</div>
<div id="ref-Sonabend2022a" class="csl-entry" role="listitem">
Sonabend, Raphael, Florian Pfisterer, Alan Mishler, Moritz Schauer, Lukas Burk, Sumantrak Mukherjee, and Sebastian Vollmer. 2022. <span>“<span class="nocase">Flexible Group Fairness Metrics for Survival Analysis</span>,”</span> May. <a href="https://doi.org/10.48550/arxiv.2206.03256">https://doi.org/10.48550/arxiv.2206.03256</a>.
</div>
<div id="ref-Sparapani2016" class="csl-entry" role="listitem">
Sparapani, Rodney A., Brent R. Logan, Robert E. McCulloch, and Purushottam W. Laud. 2016. <span>“<span class="nocase">Nonparametric survival analysis using Bayesian Additive Regression Trees (BART)</span>.”</span> <em>Statistics in Medicine</em> 35 (16): 2741–53. <a href="https://doi.org/10.1002/SIM.6893">https://doi.org/10.1002/SIM.6893</a>.
</div>
<div id="ref-Sparapani2021" class="csl-entry" role="listitem">
Sparapani, Rodney, Charles Spanbauer, and Robert McCulloch. 2021. <span>“<span class="nocase">Nonparametric Machine Learning and Efficient Computation with Bayesian Additive Regression Trees: The BART R Package</span>.”</span> <em>Journal of Statistical Software</em> 97 (1): 1–66. <a href="https://doi.org/10.18637/JSS.V097.I01">https://doi.org/10.18637/JSS.V097.I01</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/mlr-org\.com\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Created with <i class="bi bi-heart-fill"></i> for #rstats.</p>
</div>   
    <div class="nav-footer-center">
<p>CC BY SA 4.0</p>
</div>
    <div class="nav-footer-right">
<p>Website powered by <a href="https://quarto.org/">quarto</a> and <a href="https://www.netlify.com/">Netlify</a>.</p>
</div>
  </div>
</footer>




</body></html>